internal/federationin/federationin.go

```diff
@@ -232,100 +298,131 @@ func pull(ctx context.Context, metrics metrics.Exporter, opts *pullOptions) (err
-		for _, ctr := range response.Response {
-			var upperRegions []string
-			for _, region := range ctr.RegionIdentifiers {
-				upperRegions = append(upperRegions, strings.ToUpper(strings.TrimSpace(region)))
+				newExposures = append(newExposures, exposure)
 			}
-			sort.Strings(upperRegions)
-
-			for _, cti := range ctr.ContactTracingInfo {
-				for _, key := range cti.ExposureKeys {
-					if cti.TransmissionRisk < verifyapi.MinTransmissionRisk || cti.TransmissionRisk > verifyapi.MaxTransmissionRisk {
-						logger.Errorf("invalid transmission risk %v - dropping record.", cti.TransmissionRisk)
-						continue
-					}
-
-					exposures = append(exposures, &publishmodel.Exposure{
-						TransmissionRisk: int(cti.TransmissionRisk),
-						ExposureKey:      key.ExposureKey,
-						Regions:          upperRegions,
-						FederationSyncID: syncID,
-						IntervalNumber:   key.IntervalNumber,
-						IntervalCount:    key.IntervalCount,
-						CreatedAt:        createdAt,
-						LocalProvenance:  false,
-					})
-
-					if len(exposures) == fetchBatchSize {
-						resp, err := opts.deps.insertExposures(ctx, &publishdb.InsertAndReviseExposuresRequest{
-							Incoming: exposures,
-						})
-						if err != nil {
-							metricsMiddleWare.RecordPullInsertions(ctx, len(exposures))
-							return fmt.Errorf("inserting %d exposures: %w", len(exposures), err)
-						}
-						total += int(resp.Inserted)
-						exposures = nil // Start a new batch.
-					}
-				}
+			resp, err := opts.deps.insertExposures(ctx, &publishdb.InsertAndReviseExposuresRequest{
+				Incoming:     newExposures,
+				SkipRevions:  true,
+				RequireToken: false,
+			})
+			if err != nil {
+				return fmt.Errorf("inserting %d exposures: %w", len(newExposures), err)
 			}
+			// Success, update metrics
+			metricsMiddleWare.RecordPullInsertions(ctx, int(resp.Inserted))
+			metricsMiddleWare.RecordPullDroped(ctx, int(resp.Dropped))
+
+			total += int(resp.Inserted)
+		} else {
+			logger.Info("no primary keys in response")
 		}
-		if len(exposures) > 0 {
+
+		// Handle any revised keys.
+		if len(response.RevisedKeys) > 0 {
+			// Build state for new inserts.
+			revisedExposures := make([]*publishmodel.Exposure, 0, len(response.Keys))
+			for _, key := range response.RevisedKeys {
+				exposure := buildExposure(key)
+				// Fill in federation specific items.
+				exposure.FederationSyncID = syncID
+				exposure.FederationQueryID = opts.query.QueryID
+
+				if err := exposure.AdjustAndValidate(&transformSettings); err != nil {
+					logger.Errorw("invalid key on federation, skipping", "error", err)
+					continue
+				}
+
+				revisedExposures = append(revisedExposures, exposure)
+			}
 			resp, err := opts.deps.insertExposures(ctx, &publishdb.InsertAndReviseExposuresRequest{
-				Incoming: exposures,
+				Incoming:              revisedExposures,
+				OnlyRevisions:         true,
+				RequireToken:          false,
+				RequireQueryID:        true,
+				AllowPartialRevisions: true,
```

Comment:
Should this be true?

Reply:
I think it's actually irrelevant, the federation matching is done before revision token matching

so if a key is proposed for revision, but it was inserted by a different federation query ID, it will be skipped.